File: cpp/examples/contacts_manager/contact_form/contact_form.cpp
#include "contact_form.hpp"

#include <algorithm>
#include <cctype>
#include <string_view>
#include <utility>

namespace contacts_manager_example
{

namespace
{

constexpr std::string_view kCategoryDefault = "Client";

[[nodiscard]] std::string getOrEmpty(const ContactRecord& contact, const std::string& key)
{
    const auto it = contact.find(key);
    return it == contact.end() ? "" : it->second;
}

[[nodiscard]] std::string trim(std::string text)
{
    const auto begin = std::find_if_not(text.begin(), text.end(), [](unsigned char ch)
    {
        return std::isspace(ch);
    });
    const auto end = std::find_if_not(text.rbegin(), text.rend(), [](unsigned char ch)
    {
        return std::isspace(ch);
    }).base();
    if (begin >= end)
    {
        return {};
    }
    return std::string(begin, end);
}

}  // namespace

ContactForm::ContactForm(ContactRecord contact, ContactFormCallback onClose)
    : contact_(std::move(contact)), onClose_(std::move(onClose))
{
    loadContact(contact_);
}


void ContactForm::onSaveBtnClick()
{
    (void)submit();
}

void ContactForm::onCancelBtnClick()
{
    cancel();
}

void ContactForm::loadContact(const ContactRecord& contact)
{
    first_name->setValue(getOrEmpty(contact, "first_name"));
    last_name->setValue(getOrEmpty(contact, "last_name"));
    email->setValue(getOrEmpty(contact, "email"));
    phone->setValue(getOrEmpty(contact, "phone"));
    company->setValue(getOrEmpty(contact, "company"));
    city->setValue(getOrEmpty(contact, "city"));
    const std::string categoryValue = getOrEmpty(contact, "category").empty() ? std::string(kCategoryDefault) : getOrEmpty(contact, "category");
    const auto& options = category->options();
    const auto found = std::find(options.begin(), options.end(), categoryValue);
    category->setSelectedIndex(found == options.end() ? 0 : static_cast<int>(std::distance(options.begin(), found)));
    const std::string activeValue = getOrEmpty(contact, "active");
    active->setChecked(toBool(activeValue.empty() ? "true" : activeValue));
    notes->setValue(getOrEmpty(contact, "notes"));
}

bool ContactForm::submit()
{
    const std::string error = validationError();
    if (!error.empty())
    {
        message->setText(error);
        return false;
    }

    result_ = contact_;
    result_["first_name"] = trim(first_name->value());
    result_["last_name"] = trim(last_name->value());
    result_["email"] = trim(email->value());
    result_["phone"] = trim(phone->value());
    result_["company"] = trim(company->value());
    result_["city"] = trim(city->value());
    const auto& options = category->options();
    result_["category"] = options.empty() ? std::string(kCategoryDefault) : options[static_cast<std::size_t>(category->selectedIndex())];
    result_["active"] = active->checked() ? "true" : "false";
    result_["notes"] = trim(notes->value());

    if (onClose_)
    {
        onClose_(result_);
    }
    return true;
}

void ContactForm::cancel()
{
    result_.clear();
    if (onClose_)
    {
        onClose_({});
    }
}

std::string ContactForm::validationError() const
{
    if (trim(first_name->value()).empty())
    {
        return "First name is required.";
    }
    if (trim(last_name->value()).empty())
    {
        return "Last name is required.";
    }
    const std::string emailValue = trim(email->value());
    if (emailValue.empty())
    {
        return "Email is required.";
    }
    const std::size_t at = emailValue.find('@');
    if (at == std::string::npos || at == 0 || emailValue.find('.', at + 1) == std::string::npos)
    {
        return "Email must look like name@example.com.";
    }
    return {};
}

bool ContactForm::toBool(const std::string& value)
{
    std::string lower = value;
    std::transform(lower.begin(), lower.end(), lower.begin(), [](unsigned char ch)
    {
        return static_cast<char>(std::tolower(ch));
    });
    return lower == "1" || lower == "true" || lower == "yes" || lower == "y";
}

}  // namespace contacts_manager_example

File: go/examples/contacts_manager/contact_form/contact_form.go
package contact_form

import (
	"strings"
)

const contactFormDefaultCategory = "Client"

type ContactRecord map[string]string

type ContactForm struct {
	*ContactFormUI
	contact ContactRecord
	result  ContactRecord
	onClose func(ContactRecord)
}

func NewContactForm(contact ContactRecord, onClose func(ContactRecord)) *ContactForm {
	form := &ContactForm{
		ContactFormUI: NewContactFormUI(),
		contact:       copyContactRecord(contact),
		onClose:       onClose,
	}
	form.SetEventHandler(form)
	form.LoadContact(contact)
	return form
}

func (form *ContactForm) OnSaveBtnClick() {
	form.Submit()
}

func (form *ContactForm) OnCancelBtnClick() {
	form.Cancel()
}

func (form *ContactForm) LoadContact(contact ContactRecord) {
	form.FirstName.SetValue(contactValue(contact, "first_name"))
	form.LastName.SetValue(contactValue(contact, "last_name"))
	form.Email.SetValue(contactValue(contact, "email"))
	form.Phone.SetValue(contactValue(contact, "phone"))
	form.Company.SetValue(contactValue(contact, "company"))
	form.City.SetValue(contactValue(contact, "city"))
	categoryValue := contactValue(contact, "category")
	if categoryValue == "" {
		categoryValue = contactFormDefaultCategory
	}
	form.Category.SetSelectedIndex(indexOfString(form.Category.Options, categoryValue))
	activeValue := contactValue(contact, "active")
	if activeValue == "" {
		activeValue = "true"
	}
	form.Active.SetChecked(contactBool(activeValue))
	form.Notes.SetValue(contactValue(contact, "notes"))
}

func (form *ContactForm) Submit() bool {
	if message := form.validationError(); message != "" {
		form.Message.SetText(message)
		return false
	}
	result := copyContactRecord(form.contact)
	result["first_name"] = strings.TrimSpace(form.FirstName.Value)
	result["last_name"] = strings.TrimSpace(form.LastName.Value)
	result["email"] = strings.TrimSpace(form.Email.Value)
	result["phone"] = strings.TrimSpace(form.Phone.Value)
	result["company"] = strings.TrimSpace(form.Company.Value)
	result["city"] = strings.TrimSpace(form.City.Value)
	result["category"] = contactFormDefaultCategory
	if len(form.Category.Options) > 0 && form.Category.SelectedIndex >= 0 && form.Category.SelectedIndex < len(form.Category.Options) {
		result["category"] = form.Category.Options[form.Category.SelectedIndex]
	}
	result["active"] = "false"
	if form.Active.Checked {
		result["active"] = "true"
	}
	result["notes"] = strings.TrimSpace(form.Notes.Value)
	form.result = result
	if form.onClose != nil {
		form.onClose(result)
	}
	return true
}

func (form *ContactForm) Cancel() {
	form.result = ContactRecord{}
	if form.onClose != nil {
		form.onClose(ContactRecord{})
	}
}

func (form *ContactForm) Result() ContactRecord {
	return copyContactRecord(form.result)
}

func (form *ContactForm) validationError() string {
	if strings.TrimSpace(form.FirstName.Value) == "" {
		return "First name is required."
	}
	if strings.TrimSpace(form.LastName.Value) == "" {
		return "Last name is required."
	}
	email := strings.TrimSpace(form.Email.Value)
	if email == "" {
		return "Email is required."
	}
	at := strings.Index(email, "@")
	if at <= 0 || !strings.Contains(email[at+1:], ".") {
		return "Email must look like name@example.com."
	}
	return ""
}

func contactValue(contact ContactRecord, key string) string {
	if contact == nil {
		return ""
	}
	return contact[key]
}

func copyContactRecord(contact ContactRecord) ContactRecord {
	result := ContactRecord{}
	for key, value := range contact {
		result[key] = value
	}
	return result
}

func indexOfString(values []string, needle string) int {
	for index, value := range values {
		if value == needle {
			return index
		}
	}
	return 0
}

func contactBool(value string) bool {
	switch strings.ToLower(value) {
	case "1", "true", "yes", "y":
		return true
	default:
		return false
	}
}
